teachertool/src/components/makecodeFrame.tsx

```diff
@@ -0,0 +1,170 @@
+
+
+interface MakeCodeFrameProps {
+    pageSourceUrl: string;
+    highContrast?: boolean;
+    onFrameOpen: () => void;
+    onFrameClose: () => void;
+    tutorialEventHandler?: (event: pxt.editor.EditorMessageTutorialEventRequest) => void;
+    // TODO: make it so the iframe can work with a redux store
+}
+
+type FrameState = "loading" | "no-project" | "opening-project" | "project-open" | "closing-project";
+
+interface PendingMessage {
+    original: pxt.editor.EditorMessageRequest;
+    handler: (response: any) => void;
+}
+
+export const MakeCodeFrame: React.FC<MakeCodeFrameProps> =
+    ( { pageSourceUrl,
+        highContrast,
+        onFrameClose,
+        onFrameOpen
+    } ) => {
+    let ref: HTMLIFrameElement | undefined;
+    const messageQueue: pxt.editor.EditorMessageRequest[] = [];
+    let nextId: number = 0;
+    let pendingMessages: {[index: string]: PendingMessage} = {};
+    const [frameState, setFrameState] = useState<FrameState>("loading");
+    const [loadPercent, setLoadPercent] = useState(0);
+    const [workspaceReady, setWorkspaceReady] = useState(false);
+
+    let frameRef = useRef<HTMLIFrameElement>(null);
+
+    useEffect(() => {
+        // logic we want to do when the iframe is loaded
+        const root = document.getElementById("root");
+        if (ref && ref.contentWindow) {
+            window.addEventListener("message", onMessageReceived);
+            ref.addEventListener("load", handleFrameReload)
+
+            if (root) pxt.BrowserUtils.addClass(root, "editor");
+        }
+
+        // logic we want when the iframe unmounts
+        return () => {
+            window.removeEventListener("message", onMessageReceived);
+            if (root) pxt.BrowserUtils.removeClass(root, "editor");
+        }
+    }, [])
+
+    useEffect(() => {
+        sendMessageAsync({
+            type: "pxteditor",
+            action: "sethighcontrast",
+            on: highContrast
+        }  as pxt.editor.EditorMessageSetHighContrastRequest);
+    }, [highContrast])
+
+
+    useEffect(() => {
+        if (frameState === "project-open" /*&& this.props.save*/) {
+            setFrameState("closing-project");
+            onFrameClose();
+        }
+        else if (frameState === "no-project") {
+            setFrameState("opening-project");
+            onFrameOpen();
+        }
+    }, [frameState]);
+
+
+    const handleFrameReload = () => {
+        setFrameState("loading")
+    }
+
+    const onMessageReceived = (event: MessageEvent) => {
+        const data = event.data as pxt.editor.EditorMessageRequest;
+        if (frameState === "opening-project") setLoadPercent(Math.min((loadPercent || 0) + 7, 95));
+
+        if (data.type === "pxteditor" && data.id && pendingMessages[data.id]) {
+            const pending = pendingMessages[data.id];
+            pending.handler(data);
+            delete pendingMessages[data.id];
+            return;
+        }
+
+        switch (data.action) {
```

Comment:
Similar to removing the commented-out code, I'd recommend taking the unimplemented sections of this switch statement and moving them to a different branch for now. I would uncomment and keep the default statement though, which prints out the message. I think that could be quite helpful.

Reply:
Will do. As I said in your other comment about removing the comments, I don't think it's necessary to put this in a separate branch. The skillmap iframe holds a good reference for everything here that had no-ops, so we can look there.